codegen/type_converter_test.go

```diff
@@ -2530,3 +2707,553 @@ func TestCoverIsTransform(t *testing.T) {
+				if in != nil {
+					out = &structs.Bar{}
+					out.Three = string(in.Three)
+					out.Recur = convertBarHelper(in.Recur)
+				} else {
+					out = nil
+				}
+				return
+			}
+			out.Recur = convertBarHelper(in.Recur)
+		}
+	`), lines)
+}
+
+func TestConverterRecursiveNested(t *testing.T) {
+	fieldMap := make(map[string]codegen.FieldMapperEntry)
+
+	lines, err := convertTypes(
+		"Foo", "Bar",
+		`
+		struct StringFoo {
+			1: required string one
+			2: optional StringFoo recur 
+		}
+
+		struct Foo {
+			1: required string three
+			2: required StringFoo recur
+		}
+
+		struct Bar {
+			1: required string three
+			2: required StringFoo recur
+		}`,
+		nil,
+		fieldMap,
+	)
+
+	assert.NoError(t, err)
+
+	assertPrettyEqual(t, trim(`
+		out.Three = string(in.Three)
+		{
+			var convertStringFooHelper func(in *structs.StringFoo) (out *structs.StringFoo)
+			convertStringFooHelper = func(in *structs.StringFoo) (out *structs.StringFoo) {
+				if in != nil {
+					out = &structs.StringFoo{}
+					out.One = string(in.One)
+					out.Recur = convertStringFooHelper(in.Recur)
+				} else {
+					out = nil
+				}
+				return
+			}
+			out.Recur = convertStringFooHelper(in.Recur)
+		}
+	`), lines)
+}
+
+func TestConverterRecursiveMixedDeep(t *testing.T) {
+	fieldMap := make(map[string]codegen.FieldMapperEntry)
+
+	lines, err := convertTypes(
+		"Foo", "Bar",
+		`
+		struct DeepFoo {
+			1: required string ten
+			2: optional DeepFoo recur
+		}
+
+		struct StringFooA {
+			1: required string one
+			2: optional StringFooA recur 
+			3: required DeepFoo deep
+		}
+
+		struct StringFooB {
+			1: required string one
+			2: optional StringFooB recur 
+			3: required DeepFoo deep
+		}
+
+		struct Foo {
+			1: required string three
+			2: required StringFooA recur
+			3: optional Foo recur2
+		}
+
+		struct Bar {
+			1: required string three
+			2: required StringFooB recur
+			3: optional Bar recur2
+		}`,
+		nil,
+		fieldMap,
+	)
+
+	assert.NoError(t, err)
+
+	assertPrettyEqual(t, trim(`
+		out.Three = string(in.Three)
+		{
+			var convertStringFooBHelper func(in *structs.StringFooA) (out *structs.StringFooB)
+			convertStringFooBHelper = func(in *structs.StringFooA) (out *structs.StringFooB) {
+				if in != nil {
+					out = &structs.StringFooB{}
+					out.One = string(in.One)
+					out.Recur = convertStringFooBHelper(in.Recur)
+					{
+						var convertDeepFooHelper func(in *structs.DeepFoo) (out *structs.DeepFoo)
+						convertDeepFooHelper = func(in *structs.DeepFoo) (out *structs.DeepFoo) {
+							if in != nil {
+								out = &structs.DeepFoo{}
+								out.Ten = string(in.Ten)
+								out.Recur = convertDeepFooHelper(in.Recur)
+							} else {
+								out = nil
+							}
+							return
+						}
+						out.Deep = convertDeepFooHelper(in.Deep)
+					}
+				} else {
+					out = nil
+				}
+				return
+			}
+			out.Recur = convertStringFooBHelper(in.Recur)
+		}
+		{
+			var convertBarHelper func(in *structs.Foo) (out *structs.Bar)
```

Comment:
consider such code on prod with 30K qps, we definitely want to know the performance, have we benchmarked this?

also I feel if we ever have any issues around generated code during runtime for type_converter we won't be able to debug it easily

Reply:
In theory, the performance difference should only come from the extra call to the function literal

Reply:
I'll change it so that when the input & output are both non-recursive, the generated code will stay exactly like before

Reply:
> I'll change it so that when the input & output are both non-recursive, the generated code will stay exactly like before

I was about to suggest the same thing, recursive structs will be rare corner cases. Having a function for a non-recursive struct is not necessary.